### ISSUE TITLE
Strange behaviour of unreachable_from_main

### ISSUE THREAD

**Issue by incendial** [dart-lang/linter#4652]
**Describe the issue**
When a part file contains a main function (which is probably not expected to happen), the lint rule shows warnings that should be displayed in the main file (the target of `part of`) in **both** files simultaneously.

<img width="1382" alt="Screenshot 2023-08-09 at 09 43 59" src="https://github.com/dart-lang/linter/assets/21116742/97c0cfa3-c376-4a49-8219-00f66caa5dfa">

Code example:
```dart
part 'other.dart';

class Some {
  static const field = 'value';

  const Some();

  void work() {
    print(field);
  }
}
```

```dart
part of 'some.dart';

const _var1 = 1;
const _var2 = 2;
const _var3 = 3;
const _var4 = 4;

void main() {
  print(_var1);
  print(_var2);
  print(_var3);
  print(_var4);
}
```

**Additional context**
Dart SDK version: 3.2.0-41.0.dev

**Comment by bwilkerson** [comment]
@srawlins

**Comment by srawlins** [comment]
Good catch!

**Comment by rrousselGit** [comment]
Any chance we could see this fixed? It's quite annoying to come across.

It's especially annoying because it forces adding some `// ignore_for_file` in files that have no business having those. Consider:

```dart
part 'other.dart';

void main() {}
```

```dart
part of 'some.dart';

// ignore: unreachable_from_main
const _var = 42;
```


Even though the lint is correctly ignored in the right place, the lint still shows up in the wrong file.

**Comment by pq** [comment]
@srawlins

**Comment by srawlins** [comment]
> Any chance we could see this fixed? It's quite annoying to come across.

Certainly. It's in the queue.

**Comment by srawlins** [comment]
Fixed with https://github.com/dart-lang/sdk/commit/6413dd1178d7fedca1747e7f3699388865aa2980
